docs/source/bottleneck.rst

```diff
@@ -0,0 +1,31 @@
+torch.utils.bottleneck
+===============
+
+.. currentmodule:: torch.utils.bottleneck
+
+`torch.utils.bottleneck` is a tool that can be used as an initial step for
+debugging bottlenecks in your program. It summarizes runs of your script with 
+the Python profiler and PyTorch's autograd profiler. 
+
+Run it on the command line with 
+
+::
+
+    python -m torch.utils.bottleneck -- /path/to/source/script.py [args]
+
+where [args] are any number of arguments to `script.py`, or run
+``python -m torch.utils.bottleneck -h`` for more usage instructions.
+
+.. warning::
+    Because your script will be profiled, please ensure that it exits in a 
+    finite amount of time.
+
+.. warning::
+    Due to the asynchronous nature of CUDA kernels, when running against
+    CUDA code, the cProfile output and CPU-mode autograd profilers may
+    not show correct timings. In this case, the CUDA-mode autograd
+    profiler is better at assigning blame to the relevant operator(s).
```

Comment:
The picture for CUDA is not this easy, because what you should really do would be to check if your script is CPU-bound ("CPU time total is much smaller than CUDA time total" is a fairly good heuristic - profiling overhead is invisible if CPU is far ahead of the GPU anyway). Only if this is not the case should you start looking for responsible CUDA operators.

Of course the reality is much more complicated and your script can be in any of those two regimes depending on the part of the model you're evaluating. If the previous steps don't help you could try `nvprof` with `emit_nvtx`.

---

It would be nice to expand on this like I did above

Reply:
Another important thing - profiler will include the startup time (CUDA buffer allocation cost!) if you run it like this

Reply:
Sorry, two quick questions:

> "CPU time total is much smaller than CUDA time total" is a fairly good heuristic - profiling overhead is invisible if CPU is far ahead of the GPU anyway

By profiling overhead, do you mean the overhead of profiling CUDA ops or in general?

How does the CPU time total being smaller than the CUDA time total imply that the script is CPU-bound? Shouldn't it be the other way around?

Reply:
I'm not sure how to answer the first question. Most of the cost comes from profiling of CUDA ops so you have that, but what matters is that if CPU is running way ahead of the GPU then the (general) profiling costs will be hidden by asynchronous execution.

Yes, you're right of course! It's the other way around 😊 You're CPU bound if GPU time is much smaller or is close to CPU time.

Reply:
Thank you for the response, that makes more sense now. I'll send a followup PR to to update the docs soon.